src/lib/data/attachment.ts

```diff
@@ -0,0 +1,344 @@
+      if (!parentEntity) {
+        throw new Error(
+          `Internal data inconsistency found for planEntity ${parent.id}`
+        );
+      }
+      if (parentEntity.governingEntity) {
+        governingEntity = parentEntity.governingEntity;
+      }
+      supportsEntities.push(
+        ...calculateReflectiveTransitiveEntitySupport({
+          planEntities,
+          planEntity: parentEntity,
+        })
+      );
+    }
+    const data = typeCheckAttachmentData({
+      attachment,
+      attachmentVersion,
+      log,
+    });
+    result.set(attachment.id, {
+      id: attachment.id,
+      customRef,
+      parent,
+      governingEntity,
+      supportsEntities,
+      data,
+    });
+  }
+
+  return result;
+};
+
+/**
+ * Calculate a type-safe representation of the parent of an attachment,
+ * and also validate the ID against a map of allowed entities.
+ */
+const computePlanAttachmentParent = ({
+  attachment,
+  planEntities,
+  governingEntities,
+}: {
+  attachment: InstanceDataOfModel<Database['attachment']>;
+  /**
+   * A map of plan entities, which **must** include any entity that is the
+   * parent of the given attachment.
+   *
+   * Usually all the entities for a given plan are provided
+   */
+  planEntities: ValidatedPlanEntities;
+  /**
+   * A map of governing entities, which **must** include any governing entity
+   * that is the parent of the given attachment.
+   *
+   * Usually all the governing entities for a given plan are provided
+   */
+  governingEntities: MapOfGoverningEntities;
+}): PlanAttachmentParent => {
+  if (attachment.objectType === 'plan') {
+    return { type: 'plan' };
+  } else if (attachment.objectType === 'governingEntity') {
+    const ge = governingEntities.get(createBrandedValue(attachment.objectId));
+    if (!ge) {
+      throw new Error(
+        `Couldn't find governingEntity for attachment ${attachment.id}`
+      );
+    }
+    return { type: 'governingEntity', id: ge.governingEntity.id };
+  } else if (attachment.objectType === 'planEntity') {
+    const e = planEntities.get(createBrandedValue(attachment.objectId));
+    if (!e) {
+      throw new Error(
+        `Couldn't find planEntity for attachment ${attachment.id}`
+      );
+    }
+    return { type: 'planEntity', id: e.id };
+  } else {
+    throw new Error(`Invalid objectType for attachment ${attachment.id}`);
+  }
+};
+
+/**
+ * Type-Check the given attachment's value against its type, and return both the
+ * type and the value in a type-safe manner.
+ */
+export const typeCheckAttachmentData = ({
+  attachment,
+  attachmentVersion,
+  log,
+}: {
+  attachment: InstanceDataOfModel<Database['attachment']>;
+  attachmentVersion: InstanceDataOfModel<Database['attachmentVersion']>;
+  log: SharedLogContext;
+}): PlanAttachmentData => {
+  const result = {
+    type: attachment.type,
+    value: attachmentVersion.value,
+  };
+  const decodedValue = ATTACHMENT_DATA.decode(result);
+  if (!isRight(decodedValue)) {
```

Comment:
`isLeft` doesn't get any love :smile:

Reply:
haha, you're right reduced nnegation would be good... I typically stick with `isRight` because "right" also means "correct" in this case (i think likely by design haha), but let's switch this out because why not, it's clear enough still.